proposals/0012-introduce-reactNativeMetadata.md

```diff
@@ -0,0 +1,403 @@
+* Whether the New Architecture is enabled or not.
+* Which JS Engine should run in the app.
+This will not be the place where we can add any kind of configuration.
+
+Practically, this RFC proposes to add a `reactNativeManifest` section to the `package.json` file of a React Native app or library.
+This new section will allow developers to follow a declarative approach to express capabilities of apps and libraries in a more formalized manner.
+
+## Rationale
+
+The need to specify **capabilities and metadata** for a react-native library is inherent to the React Native ecosystem. This need is so evident that various frameworks provided their own ways to specify those values. Look at `react-native.config.js` for the CLI or `app.json` for Expo.
+
+Another issue is that, as of today, there is no simple programmatic way to know if a NPM package is a React Native library.
+We have some heuristics, though:
+* Presence of a `react-native:*` dependency in the `peerDependencies` section of the `package.json` ([example](https://github.com/RonRadtke/react-native-blob-util/blob/80628d418a5c81439dc2c1a1f05fae6406f0ba7f/package.json#L46C10-L49))
+* Presence of React Native specific files such as `react-native.config.js` files
+Those indicators are all non-exhaustive.
+
+Similarly, there is no easy way to understand if a library is exposing a native module, a native component or not.
+The lack of this information pushed the community to create workarounds to understand that. For Example, the React Native CLI tries to circumvent it by inspecting the native code and searching for subclasses of `*ReactPackage` using complex regular expressions ([example](https://github.com/react-native-community/cli/blob/605c542d18efcb02f217d3c85726fa73a79054c2/packages/cli-platform-android/src/config/findPackageClassName.ts#L35)) due to the lack of a declarative way to express this information.
+
+Another practical example of this need is the [React Native Directory](https://reactnative.directory/) attempting to [programmatically infer](https://github.com/react-native-community/directory/pull/870) the New Architecture support from the presence of the `codegenConfig` fields in the `package.json` of a library. The `codegenConfig` field is not intended to carry semantic information about New Architecture support, and is definitely a non-exhaustive solution for this problem.
+
+While we can continue to build tooling that tries to infer metadata and capabilities of a library, all these approach would never be precise and could provide misleading information.
+
+We suggest to provide a declarative way to describe such capabilities.
+This approach would yield several benefits:
+* **predictability**: developers can easily understand what a library is capable of doing and if such capability matches their app.
+* **tooling**: tooling can easily consume this information and provide a better experience to developers.
```

Comment:
I think that one thing to modify here is to be more explicity that we want to leverage it to help people get to the new architecture more easily; after all this whole effort was started to be able to get align-deps to help developers more easily figure out if the libs they use support the new arch -> https://github.com/microsoft/rnx-kit/issues/1863

I know that we mention in the Use Cases section and also in the Proposed Tooling more in details, but I'd call it out more explicitly in either the summary or here too with a quick, clear phrase, since it's one of the main selling points.